src/proxy/http/balance.rs

```diff
@@ -118,19 +132,45 @@ where
 impl<F, A, B> Future for MakeSvc<F, A, B>
 where
     F: Future,
-    F::Item: Discover,
+    F::Item: Discover + HasEndpointStatus,
     <F::Item as Discover>::Service: svc::Service<http::Request<A>, Response = http::Response<B>>,
     A: Payload,
     B: Payload,
 {
-    type Item = Balance<WithPeakEwma<F::Item, PendingUntilFirstData>, PowerOfTwoChoices>;
+    type Item = Service<Balance<WithPeakEwma<F::Item, PendingUntilFirstData>, PowerOfTwoChoices>>;
     type Error = F::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         let discover = try_ready!(self.inner.poll());
+        let status = discover.endpoint_status();
         let instrument = PendingUntilFirstData::default();
         let loaded = WithPeakEwma::new(discover, self.default_rtt, self.decay, instrument);
-        Ok(Balance::p2c(loaded).into())
+        let balance = Balance::p2c(loaded);
+        Ok(Async::Ready(Service { balance, status }))
+    }
+}
+
+impl<S, A, B> svc::Service<http::Request<A>> for Service<S>
+where
+    S: svc::Service<http::Request<A>, Response = http::Response<B>, Error = proxy::Error>,
+{
+    type Response = http::Response<B>;
+    type Error = fallback::Error<A>;
+    type Future = future::Either<
+        future::MapErr<S::Future, fn(proxy::Error) -> Self::Error>,
+        future::FutureResult<http::Response<B>, fallback::Error<A>>,
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.balance.poll_ready().map_err(From::from)
+    }
+
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
+        if self.status.is_empty() {
```

Comment:
Where is this set?

Reply:
Oh, I see, this is configured in the constructor.

Reply:
My main question about this... it _feels_ like something we should be polling, rather than switching on. But, given that this will be driven by balance.poll_ready, it's probably fine as it is.